ml_tools/feature_generator.py

```diff
@@ -0,0 +1,362 @@
+import pandas as pd
+from uszipcode import SearchEngine
+from abc import ABC, abstractmethod
+
+"""Module for generating, transforming and aggregating features of datasets.
+  
+The ml_tools module is a tool for transforming, aggregating and generating new features from existing features
+in input datasets. Users can use built in generators or define their own generating functions and have them
+incorporated into the API. 
+  
+  Typical usage examples:
+
+  feature_generator = GeneratorClass()
+  new_data = feature_generator.generate_feature(data)
+  
+  data_aggregator = AggregatorClass(data, relationship key)
+  aggregated_data = data_aggregator.aggregate()
+  
+  user_defined_generator = custom_generator(user_defined_function)
+"""
+
+
+class FeatureGenerator(ABC):
+    """Abstract base class for all feature generators and transformers
+
+    Attributes:
+        name: name of transformation or generator (string)
+        feature_type: type of feature being generated (string)
+    """
+
+    @property
+    def name(self):
+        raise NotImplementedError('Name property is not implemented')
+
+    @property
+    def feature_type(self):
+        raise NotImplementedError('feature_type property is not implemented')
+
+    @abstractmethod
+    def generate_feature(self, data, column=None, **kwargs):
+        """Abstract method for creating and transforming features
+
+        :param data: Pandas dataframe
+        :param column: Column on which to perform operation
+        :param kwargs: additional parameters for specific implementations
+        :return: Pandas dataframe with new or transformed features
+        """
+        pass
+
+
+class Hour(FeatureGenerator):
+    """Transforms datetime data to hour of the day"""
+
+    @property
+    def name(self):
+        return 'Hour'
+
+    @property
+    def feature_type(self):
+        return 'transformation'
+
+    @classmethod
+    def generate_feature(cls, data, column=None, **kwargs):
+        """Implements generate feature
+
+        Converts all datetime data to hours or, if given a column, converts a single column to hours
+
+        :param data: input dataframe
+        :param column: column with timestamp data
+        :param kwargs: ignored
```

Comment:
Do you need **kwargs then if its ignored?

Reply:
**kwargs is necessary in case a user defined function needs parameters.